GH-20: Support explicit issuer for New-ISHSession and Test-ISHSession

Sarafian opened the pull request.

Sarafian commented:
Cleaned up some code that was commited by mistake

ddemeyer commented on Source/ISHRemote/Trisoft.ISHRemote/Cmdlets/TrisoftCmdlet.cs:
```diff
@@ -35,7 +35,7 @@ namespace Trisoft.ISHRemote.Cmdlets
     /// Progress Record: http://community.bartdesmet.net/blogs/bart/archive/2006/11/26/PowerShell-_2D00_-A-cmdlet-that-reports-progress-_2D00_-A-simple-file-downloader-cmdlet.aspx
     /// </summary>
-    public abstract class TrisoftCmdlet : Cmdlet
+    public abstract class TrisoftCmdlet : PSCmdlet
```
I rather stay with Cmdlet. Switching to PSCmdlet means PowerShell runtime dependencies come in. So the idea of using the DLL as library is affected as you need a PowerShell runtime. The PSCmdlet to Cmdlet switch was done on 20160202 as part of TS-10630, so this would be a rollback (can't find argumentation why it was done back then besides the above).
Probably means switching the above ParameterSet.Name to an actual property check on IssuerUrl and MexUrl. Not a good enough reason to introduce PSCmdlet over Cmdlet as base class.

Sarafian replied:
I understand the reasoning, but not agree. Having the ability to use the parameter set name to drive a switch condition offers great power in code. Otherwise it makes double checking all parameters very expensive. Switching to `PSCmdlet` is more future proof and if necessary a dll can be separated into a separated isolated project. Is anyone using the the binary assembly as a dll? I notice you are worried about in another comment. How valid is the concern?

ddemeyer commented on Source/ISHRemote/Trisoft.ISHRemote/Objects/Public/IshSession.cs:
```diff
@@ -80,54 +83,59 @@ public class IshSession : IDisposable
         /// Creates a session object holding contracts and proxies to the web services API. Takes care of username/password and 'Active Directory' authentication (NetworkCredential) to the Secure Token Service.
         /// </summary>
         /// <param name="webServicesBaseUrl">The url to the web service API. For example 'https://example.com/ISHWS/'</param>
-        public IshSession(ILogger logger, string webServicesBaseUrl)
```
Why did you remove 'IshSession(ILogger logger, string webServicesBaseUrl)' overload, as it essentially breaks DLL compatibility.

Sarafian replied:
Because it wasn't used. Is anyone using the dll as a class library?

ddemeyer commented on Source/ISHRemote/Trisoft.ISHRemote/Objects/Public/IshSession.cs:
```diff
@@ -80,54 +83,59 @@ public class IshSession : IDisposable
             // webServicesBaseUrl should have trailing slash, otherwise .NET throws unhandy "Reference to undeclared entity 'raquo'." error
-            _webServicesBaseUri = (webServicesBaseUrl.EndsWith("/")) ? new Uri(webServicesBaseUrl) : new Uri(webServicesBaseUrl+"/");
+            _webServicesBaseUri = (webServicesBaseUrl.EndsWith("/")) ? new Uri(webServicesBaseUrl) : new Uri(webServicesBaseUrl + "/");
+            _wsTrustIssuerUrl = new Uri(wsTrustIssuerUrl);
```
_wsTrustIssuerUrl should become _wsTrustIssuerUri (with an I instead of L)

Sarafian replied:
Too much copy paste. Will fix!

ddemeyer commented on Source/ISHRemote/Trisoft.ISHRemote/Objects/Public/IshSession.cs:
```diff
@@ -80,54 +83,59 @@ public class IshSession : IDisposable
-            _webServicesBaseUri = (webServicesBaseUrl.EndsWith("/")) ? new Uri(webServicesBaseUrl) : new Uri(webServicesBaseUrl+"/");
+            _webServicesBaseUri = (webServicesBaseUrl.EndsWith("/")) ? new Uri(webServicesBaseUrl) : new Uri(webServicesBaseUrl + "/");
+            _wsTrustIssuerUrl = new Uri(wsTrustIssuerUrl);
+            _wsTrustIssuerMexUrl = new Uri(wsTrustIssuerMexUrl);
```
_wsTrustIssuerMexUrl should become _wsTrustIssuerMexUri (with an I instead of L)

Sarafian replied:
Too much copy paste. Will fix!

ddemeyer commented on Source/ISHRemote/Trisoft.ISHRemote/Objects/Public/IshSession.cs:
```diff
@@ -141,13 +149,20 @@ private void CreateConnection()
             var connectionParameters = new InfoShareWcfConnectionParameters
             {
-                Credential = _ishSecurePassword == null ? new NetworkCredential() : new NetworkCredential(_ishUserName, SecureStringConversions.SecureStringToString(_ishSecurePassword)),
+                Credential = _ishSecurePassword == null ? null : new NetworkCredential(_ishUserName, SecureStringConversions.SecureStringToString(_ishSecurePassword)),
```
"? null" should be "? new NetworkCredential" ... or was that a double initialization... was this tested? Why change?

Sarafian replied:
When not providing credentials, it seems correct to retain credentials as `null` for the entire flow. It also makes a condition that decides about the binding type easier.

ddemeyer replied:
I don't think that that is tested in the Tests as it required Windows Authentication on the STS (so ADFS mostly).

Sarafian replied:
Tested windows authentication with ADFS.

Sarafian commented:
You are welcome. 

2/3 comments relate to a concern about using dll as binary assembly from e.g. a console application. I would expect that to be a nuget package for example. If the concern is still valid and has merit, I'll adapt the code. Otherwise I suggest we keep `PSCmdlet` and lose code that is not used. Regarding the 

> New NetworkCredential

please read the comment and look into the code flow. I believe the pre-existing code didn't do anything.  Besides `New-ISHSession` and `Test-ISHSession` tests that drive this part were successful.

Sarafian commented:
- Fixed the typos for the `Url` as suggested above
- Added a safe guard error to improve the error when credentials are expected but were not provided
- Didn't do anything regarding "dll compatibility" as I don't currently see it's value with this approach. Left commented code to make easy the transition if necessary.
